events_mgmt/src/main/java/com/adobe/event/management/RegistrationService.java

```diff
@@ -11,11 +11,63 @@
  */
 package com.adobe.event.management;
 
+import com.adobe.Workspace;
 import com.adobe.event.management.model.Registration;
 import com.adobe.event.management.model.RegistrationInputModel;
+import feign.RequestInterceptor;
+import java.util.Optional;
+import org.apache.commons.lang3.StringUtils;
 
 public interface RegistrationService {
 
-  Registration createRegistration(RegistrationInputModel registrationInputModel);
+  Optional<Registration> findById(String registrationId);
 
+  Optional<Registration> createRegistration(
+      RegistrationInputModel.Builder registrationInputModelBuilder);
+
+  static Builder builder() {
+    return new Builder();
+  }
+
+  class Builder {
+
+    private RequestInterceptor authInterceptor;
+    private Workspace workspace;
+    private String url;
+
+    private static void validateWorkspace(Workspace workspace) {
+      if (StringUtils.isEmpty(workspace.getImsOrgId())) {
+        throw new IllegalArgumentException("Workspace is missing an imsOrgId context");
+      }
+      if (StringUtils.isEmpty(workspace.getConsumerOrgId())) {
+        throw new IllegalArgumentException("Workspace is missing a consumerOrgId context");
+      }
+      if (StringUtils.isEmpty(workspace.getCredentialId())) {
+        throw new IllegalArgumentException("Workspace is missing a credentialId context");
+      }
+      if (StringUtils.isEmpty(workspace.getApiKey())) {
+        throw new IllegalArgumentException("Workspace is missing an apiKey context");
+      }
```

Comment:
1. This validation code is common and can be moved to the Workspace constructor for re-use. wdyt?
2. Are we following some paradigm on where to place the validation code? Is it inside Builder class or the constructor of the corresponding model class?

Reply:
1. No it's not common. Elsewhere the api-key is only needed and set by the IMS auth interceptor. The Registration API is special, it expects it in the header of  the registration input model 
2. The expected `workspace` entries vary from one service to the next, but to follow your advise, for one common validation paradigm, I moved the validation at the service constructor level, wdyt ?

Reply:
Looks good 👍